feat(profiles): Emit "accepted" outcomes for profiles filtered by sampling

jjbayer opened the pull request:
To facilitate a billing model that is consistent between transactions and profiles, #2051 introduced a new data category for profiles, such that

```
processed profiles = indexed profiles + sampled profiles  # "sampled" as in dropped by dynamic sampling
```

just like it is already the case for transactions:

```
processed transactions = indexed transactions + sampled transactions
```

In other words, "processed profiles" should count _all_ valid, non rate limited profiles seen by Relays, even if they were dropped by dynamic sampling.

## Difference between transactions and profiles

For transactions, we extract _metrics_ before dynamic sampling, and those metrics are what we rate limit and eventually log as "accepted" for the "processed" transaction data category (`DataCategory.Transaction`).

For profiles, we do not extract metrics (yet), so the outcomes for the "processed" profile data category have to be calculated in a different way.

## How this PR achieves this goal

1. In processing Relays, if an envelope still contains profiles _after_ dynamic sampling, log an `Accepted` outcome for the "processed" category. By restricting this to processing Relays, we can be sure that every profile is only counted once.
2. Also in processing Relays, shortly before reporting outcomes to kafka,
2.1. translate `Filtered` outcomes with reason `Sampled:` to an `Accepted` outcome. This counts all profiles dropped by dynamic sampling, regardless of where the dynamic sampling took place (external relay, pop relay, processing relay).
2.2. Also send the original `Filtered` outcome, but with data category `DataCategory.TransactionIndexed`.

By adding up the counts of these two disjoint sets, we should correctly count all profiles regardless of whether they were sampled or not.

## Alternative proposal (rejected for now)

In order to _actually_ line up behavior of transactions and profiles, we could start extracting a simple counter metric for profiles before dynamic sampling, and let that metric represent `DataCategory.Profile` -- this would mean that rate limits are applied to the metric, and the accepted outcome would be emitted from the `billing_metrics_consumer`, just like we do for the `DataCategory.Transactions`. See [this internal doc](https://www.notion.so/sentry/Implementation-Concerns-412caf18c2f04f579bb551b98c9dad8c) for more.

## Why the currrent approach was chosen

I personally believe that the alternative proposal described above would be more correct and easier to maintain, but:

1. By containing all new logic to processing relays, we will correctly count "processed" profiles even if they were dropped by dynamic sampling outdated external relays.
3. In a wider sense, by containing all new logic to processing relays, we can iterate faster (deploy hotfixes, etc.) without having to worry about the behavior of external relays (which we cannot update).
4. This change is easy to revert if needed. The alternative solution would be distributed across nodes, from external relays to sentry.

jjbayer committed:
feat: Transform outcomes

jjbayer committed:
feat: Log Accepted outcome for processed profiles

jjbayer committed:
test: stub

jjbayer committed:
wip

jjbayer committed:
test

jjbayer committed:
test

jjbayer committed:
fix: Envelope summary problem

jjbayer committed:
test: passes for single relay

jjbayer committed:
test: pass

jjbayer committed:
self review

phacops commented on relay-server/src/actors/processor.rs:
```diff
@@ -1070,6 +1070,51 @@ impl EnvelopeProcessorService {
+            .items()
+            .filter(|item| item.ty() == &ItemType::Profile)
+            .map(|item| item.quantity())
+            .sum();
```
Technically, the SDKs will only send 1 profile per envelope.

jjbayer replied:
Makes sense, but Relay does not enforce that AFAIK, so its safer to handle the case where there's multiple profiles.

phacops reviewed (commented):
LGTM

olksdr reviewed (approved):
lgtm. Let's test it in prod 😄

olksdr commented on relay-server/src/actors/processor.rs:
```diff
@@ -1070,6 +1070,51 @@ impl EnvelopeProcessorService {
+    ///
+    /// In the future, we might actually extract metrics from profiles before dynamic sampling,
+    /// like we do for transactions. At that point, this code should be removed, and we should
+    /// enforce rate limits and emit outcomes based on the collect profile metric, as we do for
```
The sentence ends here kind of leaving you hanging at the end.
```suggestion
    /// enforce rate limits and emit outcomes based on the collect profile metric, as we do for transactions
```

jjbayer replied:
Thanks!

jjbayer committed:
doc: missing word

jjbayer committed:
fix(profiles): Log ProfileIndexed outcome for invalid profiles (#2056)

Improvement to #2054:

Before a profile reaches the `process_profile` stage, we count it as
`Profile:Accepted`. So if `process_profile` _fails_ with an `Invalid`
outcome, it is incorrect to log it in the same data category, because it
would double count within that category. Log it as `ProfileIndexed`
instead.

jjbayer committed:
fix(profiles): Log sampling outcome as ProfileIndexed (#2061)

Improvement to https://github.com/getsentry/relay/pull/2054:

Since #2054 already introduced a translation from `Profile` to
`ProfileIndexed` for sampling outcomes, this PR does not change anything
functionally. It is merely an attempt to be consistent with #2056 and
#2060, such that we consider the `metrics_extracted` flag any time we
generate a profiling outcome.

jjbayer committed:
fix(profiles): Log ProfileIndexed outcome for rate limited profiles (#2060)

Improvement to https://github.com/getsentry/relay/pull/2054:

Same as https://github.com/getsentry/relay/pull/2056, but for rate
limited outcomes: _Before_ metrics extraction, rate limited profiles
should count as `Profile` ("processed" profile), because they have not
been counted towards accepted profiles yet. _after_ metrics extraction,
they should count as `ProfileIndexed`.

jjbayer committed:
doc: changelog
